python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+    guarnteed to work.
+    """
+
+    paramSpecs = {
+        "sklearnEstimator": {"doc": "sklearn estimator applied to each group"},
+        "keyCols": {"doc": "list of key column names", "default": ["key"]},
+        "xCol": {"doc": "input features column name", "default": "features"},
+        "outputCol": {"doc": "output column name", "default": "output"},
+        "yCol": {"doc": "optional label column name", "default": None},
+        "estimatorType": {"doc": "scikit-learn estimator type", "default": "transformer"}}
+    """`paramSpecs[paramName]` returns a parameter specification dictionary with the
+    documentation and default value for the parameter of name `paramName`"""
+
+    @keyword_only
+    def __init__(self, sklearnEstimator=None, keyCols=["key"], xCol="features",
+                 outputCol="output", yCol=None):
+        """For all instances, the ordered list of `keyCols` determine the set of groups which each
+        `sklearnEstimator` is applied to.
+
+        For every unique `keyCols` value, the remaining columns are aggregated and used to train the
+        scikit-learn estimator.
+
+        If `yCol` is specified, then this is assumed to be of `"predictor"` type, else a
+        `"transformer"`.
+
+        :param sklearnEstimator: An instance of a sklearn estimator, with parameters configured
+        as desired for each user.
+        :param keyCols: Key column names list used to group data to which models are applied, where
+        implies order lexicographical importance.
+        :param xCol: Name of column of input features used for training and
+        transformation/prediction.
+        :param yCol: Optional. Specifies name of label column for regression or classification
+        pipelines.
+
+        :raise ValueError: if `sklearnEstimator` is `None`.
+        :raise ValueError: if `sklearnEstimator` does not derive from `sklearn.base.BaseEstimator`.
+        :raise ValueError: if `keyCols` is empty.
+        :raise ValueError: if any column has the name `"estimator"` (
+        :raise ValueError: if reflection checks indicate that parameter estimator is not equipped
+        with a `fit()` method or an appropriate transformation/prediction method.
+        """
+        if sklearnEstimator is None:
+            raise ValueError("sklearnEstimator should not be unspecified or None")
+        if not isinstance(sklearnEstimator, sklearn.base.BaseEstimator):
+            raise ValueError("sklearnEstimator should be an sklearn.base.BaseEstimator")
+        if len(keyCols) == 0:
+            raise ValueError("keyCols should not be empty")
+        if "estimator" in keyCols + [xCol, yCol]:
+            raise ValueError("keyCols should not contain a column named \"estimator\"")
+
+        for paramName, paramSpec in KeyedEstimator.paramSpecs.items():
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedEstimator, self).__init__()
```

Comment:
Ok I understand why. Please put a comment explaining why you need to depart from python conventions, it is non-obvious that the base class performs reflective introspection unless you know the pyspark code.